staging/src/k8s.io/apiserver/pkg/server/routes/openstack.go

```diff
@@ -110,21 +162,73 @@ func (o Openstack) genericOpenStackRequestHandler(target string, resp http.Respo
 	resp.Write(body)
 }
 
+func (o Openstack) actionHandler(resp http.ResponseWriter, req *http.Request, redirectUrlPath string) {
+	klog.V(3).Infof("Redirect request to %s", redirectUrlPath)
+	http.Redirect(resp, req, redirectUrlPath, http.StatusTemporaryRedirect)
+}
+
 // TODO: redirect could introduce perf impact to the server, so setup the routes for Openstack requests
 //       in the API installation flow at the api server init state will be a better solution
+// GET a server : GET - /servers/serverName
+// LIST severs:   GET - /servers
+// LIST servers:  GET - /servers, with label selector in query
+// CREATE a server: POST - /servers
+// CREATE batch servers: POST  - /servers
+// DELETE a server : DELETE - /servers/serverName
+// note that NO batch delete, in openstack
+// TODO: for post 130, we should support cases to delete all in batch if this is a valid use case
+
+// actions:
+// CREATE an action: POST - /servers/serverName/action
+//
 func (o Openstack) serverHandler(resp http.ResponseWriter, req *http.Request) {
 	klog.V(4).Infof("handle /servers. URL path: %s", req.URL.Path)
 
 	tenant := openstack.GetTenantFromRequest(req)
 	namespace := openstack.GetNamespaceFromRequest(req)
 
-	redirectUrl := fmt.Sprintf(POD_URL_TEMPLATE, tenant, namespace)
+	redirectUrl := ""
 
 	if openstack.IsActionRequest(req.URL.Path) {
+		redirectUrl = fmt.Sprintf(POD_URL_TEMPLATE, tenant, namespace)
 		redirectUrl += "/" + getElementFromPath(req.URL.Path) + "/action"
-	} else if req.Method == "GET" || req.Method == "DELETE" {
-		//Get the VM ID for redirect
+		o.actionHandler(resp, req, redirectUrl)
+		return
+	}
+
+	switch req.Method {
+	case http.MethodGet:
+		redirectUrl = fmt.Sprintf(POD_URL_TEMPLATE, tenant, namespace)
+		if isGetDetail(req.URL.Path) {
+			redirectUrl += "/" + getElementFromPath(req.URL.Path)
+		}
+
+		rev_id, err := peekListFilters(req)
+		if err != nil {
+			resp.WriteHeader(http.StatusBadRequest)
+			return
+		}
+		if rev_id != "" {
+			redirectUrl += fmt.Sprintf("?labelSelector=ln=%s", rev_id)
```

Comment:
This looks weird "labelSelector=ln=". Have you tested it?
If it works, can you add a sample URL as a comment?

Reply:
yes, this is tested as described in the PR.